### ISSUE TITLE
Validation Layer Reports Invalid attribAddress alignment for vertex attribute

### ISSUE THREAD

**Issue by youss2017** [KhronosGroup/Vulkan-ValidationLayers#3733]
I do not understand what the error is saying because there is no attribAddress alignment requirement in vkPhysicalDeviceLimits, and the spec is very vague and does not saying anything useful. This error occurs for all three attributes.

Here is the full error:
Validation Error: `[ VUID-vkCmdDrawIndexed-None-02721 ] Object 0: handle = 0xfef35a00000000a0, type = VK_OBJECT_TYPE_BUFFER; Object 1: handle = 0xa56ac00000000d4, type = VK_OBJECT_TYPE_PIPELINE; | MessageID = 0x24afafc5 | vkCmdDrawIndexed: Invalid attribAddress alignment for vertex attribute 0, VK_FORMAT_R32G32B32_SFLOAT,from of VkPipeline 0xa56ac00000000d4[] and vertex VkBuffer 0xfef35a00000000a0[]. The Vulkan spec states: For a given vertex buffer binding, any attribute data fetched must be entirely contained within the corresponding vertex buffer binding, as described in Vertex Input Description (https://vulkan.lunarg.com/doc/view/1.2.198.1/windows/1.2-extensions/vkspec.html#VUID-vkCmdDrawIndexed-None-02721)`

Here is my shader code:
```
#version 450 core

layout (location = 0) in vec3 inPosition;
layout (location = 1) in vec3 inNormal;
layout (location = 2) in vec2 inTexCoord;

// Instancing Data
//layout (location = 3) in vec2 XYOffset;

layout (binding = 0) uniform View_Projection {
	mat4 u_View;
	mat4 u_Projection;
};

layout (binding = 1) uniform Model {
	mat4 u_Model;
	mat4 u_NormalModel; // transpose(inverse(u_Model))
};

layout (location = 0) out vec3 Normal;
layout (location = 1) out vec2 TexCoord;

void main() {
	Normal = mat3(u_NormalModel) * inNormal;
	TexCoord = inTexCoord;
	//vec4 InstancePosition = vec4(inPosition.xy + XYOffset, inPosition.z, 1.0);
	//gl_Position = u_Projection * u_View * u_Model * InstancePosition;
	gl_Position = u_Projection * u_View * u_Model * vec4(inPosition, 1.0);
}
```

Here is my binding code (for vulkan)
```
struct VulkanPipelineVertexInput
{
    VkPipelineVertexInputStateCreateInfo createInfo;
    std::vector<VkVertexInputBindingDescription> binding_descriptions;
    std::vector<VkVertexInputAttributeDescription> attribute_descriptions;
    std::vector<VkVertexInputBindingDivisorDescriptionEXT> divisor_description;
    VkPipelineVertexInputDivisorStateCreateInfoEXT divisorCreateInfo;
};

static VulkanPipelineVertexInput Vulkan_Internal_PipelineState_InitalizeVertexInput(IPipelineLayout layout)
{
    VulkanPipelineVertexInput input_state;

    PipelineVertexInputDescription &input_description = layout->m_vertex_input_description;
    if(input_description.m_input_elements.size() > 0)
    {
        int lastBinding = -1;
        for(const auto& element : input_description.m_input_elements)
        {
            VkVertexInputAttributeDescription attribute;
            attribute.binding = element.m_binding_id;
            attribute.format = element.m_vk_format;
            attribute.location = element.m_location;
            attribute.offset = element.m_offset;
            input_state.attribute_descriptions.push_back(attribute);
            if(lastBinding != element.m_binding_id)
            {
                lastBinding = element.m_binding_id;
                VkVertexInputBindingDescription binding_description;
                binding_description.binding = element.m_binding_id;
                binding_description.inputRate = element.m_per_instance ? VK_VERTEX_INPUT_RATE_INSTANCE : VK_VERTEX_INPUT_RATE_VERTEX;
                binding_description.stride = element.m_stride;
                if(element.m_per_instance)
                {
                    VkVertexInputBindingDivisorDescriptionEXT divisor_description;
                    divisor_description.binding = element.m_binding_id;
                    divisor_description.divisor = element.m_divisor_rate;
                    input_state.divisor_description.push_back(divisor_description);
                }
                input_state.binding_descriptions.push_back(binding_description);
            }
        }
    }

    input_state.createInfo.sType = VK_STRUCTURE_TYPE_PIPELINE_VERTEX_INPUT_STATE_CREATE_INFO;
    input_state.createInfo.pNext = input_state.divisor_description.size() > 0 ? &input_state.divisorCreateInfo : nullptr;
    input_state.createInfo.flags = 0;
    input_state.createInfo.vertexBindingDescriptionCount = input_state.binding_descriptions.size();
    input_state.createInfo.pVertexBindingDescriptions = input_state.binding_descriptions.data();
    input_state.createInfo.vertexAttributeDescriptionCount = input_state.attribute_descriptions.size();
    input_state.createInfo.pVertexAttributeDescriptions = input_state.attribute_descriptions.data();
    return input_state;
}
```

**Comment by sfricke-samsung** [comment]
some background:

`attribAddress` is a from the [formula the spec has](https://www.khronos.org/registry/vulkan/specs/1.3-extensions/html/vkspec.html#fxvertex-input-address-calculation):

> attribAddress = bufferBindingAddress + vertexOffset + attribDesc.offset;

I think the error message really lacks giving numerical values of "why" its not aligned correct and that is the real issue here!

I am assuming (but have been surprised before) the VVL code is correct in this case due to amount of testing this VU has.

**Comment by youss2017** [comment]
What does VVL and VU stand for?
This information may be useful.
Vertex Stride: 32 bytes, Offsets 0: 0, Offsets 1: 12, Offsets 2: 24

**Comment by youss2017** [comment]
I figured it out!!!!!!!!! I thought this was a driver bug.

I opened the program in RenderDoc and it showed that the vertex buffer had an offset of 1, look:
![err](https://user-images.githubusercontent.com/30071252/153072436-323fefd1-3c9a-4452-92a0-b5765f506009.PNG)

The problem was setting the vkCmdSetVertexBuffers wrong. vkCmdSetVertexBuffers requires an array of VkDeviceSize[] pOffsets even if you don't use them. I made a typo if writing VkDeviceSize[1] pOffsets = {1} instead of VkDeviceSize[1] = {0}
